### DISCUSSION TITLE
How to set word order in the simulator

### DISCUSSION THREAD

**Discussion by Fredo70** [pymodbus-dev/pymodbus#1593] [Q&A] [answered]
I am completely new to pymodbus. I want to simulate a modbus slave with the simulator in order to test my software. The device has 32-bit integer with Little Endian. Is it possible to set the simulator like this? It seems that the uint32 are coded as Big Endian.
Thanks

**Comment by janiversen** [comment, marked as answer]
The modbus protocol is pr definition big endian, so devices that are little endian convert before sending a message.

The datastore simulator (which is used in pymodbus.simulator), does convert the configured number to big endian automatically.

If you use pymodbus in your client big endian is also done automatically.

**Reply by Fredo70** [reply]
Well. I work in building automation and I have experienced every possible combination and never anything standard ;-) But anyway.
Maybe I can describe my use case. Because I have other problems. One thing to mention is that I'm starting with Python (I'm a Java and .NET programmer), but with internet I should be able to do it.
I have a physical device (slave) and a physical display (master). In between is a data logger (from the same Chinese manufacturer). But the data logger does not work well. It crashes when the ambient temperature gets warm. Now my project is to replace this logger with a Raspberry pi. On the one hand, the pi should pass on the modbus commands (read and write) from the display 1:1 to the slave device. In addition, it should query a subset of available registers and log them in a database (e.g. SQLite).
Maybe later a websocket to control a web UI (but that is still up in the air).
Since the devices are not exactly accessible (in a caravan), my procedure was as follows:
1. configure the simulator to simulate the slave. The order is also not a big problem. Since it is only for testing, I can simply configure 2 16-bit registers.
2. read these registers with my Windows PC and a Modbus client. I have done that. It works.
3. Integrate the Modbus forwarder and connect it with the Windows PC not to the simulator but to the new server. The client then connects to the simulator. Test the forwarder functionality. Basically it works, but I get an error. I found something similar in the Issues https://github.com/pymodbus-dev/pymodbus/issues/1549#issuecomment-1563466215
But I am not writing (yet). Version is 3.3.1

```
2023-06-11 19:43:42,379 DEBUG logging:102 Factory Request[ReadInputRegistersRequest': 4]
2023-06-11 19:43:42,380 DEBUG logging:102 validate[4] 12544:4
2023-06-11 19:43:42,385 ERROR logging:114 Datastore unable to fulfill request: 'coroutine' object has no attribute 'isError'; Traceback 
(most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/pymodbus/server/async_io.py", line 264, in execute
    response = request.execute(context)
  File "/usr/local/lib/python3.9/dist-packages/pymodbus/register_read_message.py", line 209, in execute
    if not context.validate(self.function_code, self.address, self.count):
  File "/usr/local/lib/python3.9/dist-packages/pymodbus/datastore/remote.py", line 50, in validate
    return not self.result.isError()
AttributeError: 'coroutine' object has no attribute 'isError'
```
I don't know if and what I'm doing wrong.
Is my (for private use) project at all feasible with pymodbus (just a rough estimate)? Or am I completely on the wrong track?
Many thanks

**Reply by janiversen** [reply]
It sounds quite a lot like an upcoming modbus_forwarder we are working on.

The current modbus_forwarder have a problem, because it is sync. and the modbus calls are async.

What you need to do is to combine a pymodbus server with a pymodbus client. What we are looking at is to use the server callbacks instead of a datastore, truth is that the remote_datastore worked nicely for sync. but is a mess for async.

If you decide to program something, please submit as pull requests...that will enable us to help you, instead of us making the same work.

**Reply by janiversen** [reply]
I was just corrected the current idea is to use a custom framer in the server. A framer received all request and return the corresponding response, so instead of getting the response internally it will pass the request to the client and return the response from the client.

**Reply by Fredo70** [reply]
> What you need to do is to combine a pymodbus server with a pymodbus client. What we are looking at is to use the server callbacks instead of a datastore...

Thank you very much for your help. That was my next consideration. I hope I can do it, since, as I said, I am a python beginner :-) Maybe the code won't be "pretty".
I will notify you of my progress.

**Reply by Fredo70** [reply]
> I was just corrected the current idea is to use a custom framer in the server.

Ok. I will check this.
